modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/config/DynamicConfigurationInstanceProvider.java

```diff
@@ -0,0 +1,180 @@
+ * to {@link #resolverSet#resolve(MuleEvent)}, the resulting {@link ResolverSetResult} might not end up
+ * generating a new instance. This is so because {@link ResolverSetResult} instances are put in a cache to
+ * guarantee that equivalent evaluations of the {@code resolverSet} return the same instance.
+ *
+ * @since 3.7.0
+ */
+public final class DynamicConfigurationInstanceProvider<T> implements ConfigurationInstanceProvider<T>, ExpirableContainer<Object>
+{
+
+    private final String name;
+    private final Extension extension;
+    private final ConfigurationInstanceRegistrationCallback registrationCallback;
+    private final ConfigurationObjectBuilder configurationObjectBuilder;
+    private final ResolverSet resolverSet;
+    private final ExpirationPolicy expirationPolicy;
+
+    private final Map<ResolverSetResult, DynamicConfigurationInstanceHolder> cache = new ConcurrentHashMap<>();
+    private final ReadWriteLock cacheLock = new ReentrantReadWriteLock();
+    private final Lock cacheReadLock = cacheLock.readLock();
+    private final Lock cacheWriteLock = cacheLock.writeLock();
+
+    /**
+     * Creates a new instance
+     *
+     * @param configurationObjectBuilder the {@link ConfigurationObjectBuilder} that will build the configuration instances
+     * @param resolverSet                the {@link ResolverSet} that's going to be evaluated
+     */
+    public DynamicConfigurationInstanceProvider(String name,
+                                                Extension extension,
+                                                ConfigurationInstanceRegistrationCallback registrationCallback,
+                                                ConfigurationObjectBuilder configurationObjectBuilder,
+                                                ResolverSet resolverSet,
+                                                ExpirationPolicy expirationPolicy)
+    {
+        this.name = name;
+        this.extension = extension;
+        this.registrationCallback = registrationCallback;
+        this.configurationObjectBuilder = configurationObjectBuilder;
+        this.resolverSet = resolverSet;
+        this.expirationPolicy = expirationPolicy;
+    }
+
+    /**
+     * Evaluates {@link #resolverSet} using the given {@code event} and returns
+     * an instance produced with the result. For equivalent {@link ResolverSetResult}s
+     * it will return the same instance.
+     *
+     * @param operationContext a {@link OperationContext}
+     * @return the resolved value
+     */
+    @Override
+    public T get(OperationContext operationContext)
+    {
+        try
+        {
+            ResolverSetResult result = resolverSet.resolve(asOperationContextAdapter(operationContext).getEvent());
+            DynamicConfigurationInstanceHolder configurationInstanceHolder = getHolder(result);
+
+            operationContext.onOperationSuccessful(signal -> discountUsage(configurationInstanceHolder));
```

Comment:
no, because the signal is of different types. I though of giving both signals a common ancestor but it didn't make much sense